mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -0,0 +1,420 @@
+            .from(TABLE_NAME)
+            .where(MAILBOX_ID.eq((mailboxId.asUuid())))
+            .and(extraCondition)
+            .orderBy(MESSAGE_UID.asc());
+        return limit.map(limitValue -> (SelectFinalStep<Record1<Long>>) queryWithoutLimit.limit(limitValue))
+            .orElse(queryWithoutLimit);
+    }
+
+    private static final Content EMPTY_CONTENT = new Content() {
+        @Override
+        public InputStream getInputStream() {
+            return new ByteArrayInputStream(new byte[0]);
+        }
+
+        @Override
+        public long size() {
+            return 0;
+        }
+    };
+
+    private final PostgresExecutor postgresExecutor;
+
+    private final ThreadId.Factory threadIdFactory;
+
+    public PostgresMailboxMessageDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+        this.threadIdFactory = new ThreadId.Factory(new PostgresMessageId.Factory());
+    }
+
+    public Mono<MessageUid> findFirstUnseenMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                IS_SEEN.eq(false), Optional.of(1), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Flux<MessageUid> findAllRecentMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                IS_SEEN.eq(true), Optional.empty(), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Flux<MessageUid> listAllMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                DSL.noCondition(), Optional.empty(), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Mono<MessageUid> findLastMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.select(MESSAGE_UID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .orderBy(MESSAGE_UID.desc())
+                .limit(1)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Mono<Void> deleteByMailboxId(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))));
+    }
+
+    public Mono<Void> deleteByMailboxIdAndMessageUid(PostgresMailboxId mailboxId, MessageUid messageUid) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_UID.eq(messageUid.asLong()))));
```

Comment:
AND use returning CF https://github.com/apache/james-project/pull/1812/files/210221dc5627899640489bf759b985b63c87fe74..b25de17726a22392133852d03fd26a147bfd9ffd#r1405888542